core/src/util/two_pointer_traversal.rs

```diff
@@ -0,0 +1,224 @@
+) where
+    COMPARATOR: Fn(&ITEM1, &ITEM2) -> Ordering,
+    HANDLER: FnMut(&ITEM1, &ITEM2),
+{
+    // Each closure has a unique signature & so need to have these
+    // dummy ones to help the compiler
+    let extra1_handler = |_: &ITEM1| {};
+    let extra2_handler = |_: &ITEM2| {};
+    traverse(
+        list1,
+        list2,
+        comparator,
+        Some(equal_items_handler),
+        Some(extra1_handler),
+        Some(extra2_handler),
+    )
+}
+
+/// Look @ the docs for [traverse]
+pub fn diff<ITEM1, ITEM2, COMPARATOR, ExtraInList1Hander, ExtraInList2Hander>(
+    list1: &[ITEM1],
+    list2: &[ITEM2],
+    comparator: COMPARATOR,
+    extra_in_list1_handler: ExtraInList1Hander,
+    extra_in_list2_handler: ExtraInList2Hander,
+) where
+    COMPARATOR: Fn(&ITEM1, &ITEM2) -> Ordering,
+    ExtraInList1Hander: FnMut(&ITEM1),
+    ExtraInList2Hander: FnMut(&ITEM2),
+{
+    // Each closure has a unique signature & so need to have this
+    // dummy ones to help the compiler
+    let equal_handler = |_: &ITEM1, _: &ITEM2| {};
+    traverse(
+        list1,
+        list2,
+        comparator,
+        Some(equal_handler),
+        Some(extra_in_list1_handler),
+        Some(extra_in_list2_handler),
+    )
+}
+
+/// Items from list1 & list2 are compared against each other using the
+/// comparator & the handlers would be informed about the appropriate items
+///
+/// This works under a couple of assumptions
+///     - Lists are sorted
+///     - No repetitions in the lists
+fn traverse<ITEM1, ITEM2, COMPARATOR, EqualHandler, ExtraInList1Hander, ExtraInList2Hander>(
+    list1: &[ITEM1],
+    list2: &[ITEM2],
+    comparator: COMPARATOR,
+    mut equal_items_handler: Option<EqualHandler>,
+    mut extra_in_list1_handler: Option<ExtraInList1Hander>,
+    mut extra_in_list2_handler: Option<ExtraInList2Hander>,
+) where
+    COMPARATOR: Fn(&ITEM1, &ITEM2) -> Ordering,
+    EqualHandler: FnMut(&ITEM1, &ITEM2),
+    ExtraInList1Hander: FnMut(&ITEM1),
+    ExtraInList2Hander: FnMut(&ITEM2),
+{
+    let mut list1_index = 0;
+    let mut list2_index = 0;
+    let message_expected_some = "Already checked for is_some but still fails!";
+
+    // We want to exhaust both the lists in the worst case (no common items)
+    let total_items_count = list1.len() + list2.len();
+    for _ in 0..total_items_count {
+        let item1 = list1.get(list1_index);
+        let item2 = list2.get(list2_index);
+
+        // Both the lists have run out! Finish
+        if item1.is_none() && item2.is_none() {
+            break;
+        } else {
+            let ordering = if item1.is_some() && item2.is_none() {
+                Ordering::Less
+            } else if item1.is_none() && item2.is_some() {
+                Ordering::Greater
+            } else {
+                let item1 = item1.expt(message_expected_some);
+                let item2 = item2.expt(message_expected_some);
+                comparator(item1, item2)
+            };
```

Comment:
Is this the common way to write this in rust? Another possible way to write it is:
```
    let ordering = {
        if item1.is_some() && item2.is_none() {
            Ordering::Less
        } else if item1.is_none() && item2.is_some() {
            Ordering::Greater
        } else {
            let item1 = item1.expt(message_expected_some);
            let item2 = item2.expt(message_expected_some);
            comparator(item1, item2)
        };
    };
```
which I personally find more readable

Reply:
Hmmm. I actually haven't seen this done much. I just got used to putting right after the `=` since that conveys the info that it is being used as an expression to me. I'll look to see what other folks do for this